### ISSUE TITLE
Review default package_id() implementation

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#3762]
There are several quirks that should be reconsidered:

- The default package_id() uses the options of its direct dependencies, they affect the hash too. Maybe that should be removed from the default
- Consider using by default a "full_package_mode" for shared libs depending on static libs.

**Comment by danimtb** [comment]
Is this still valid regarding #3740 ?

I am sure package_id() needs a review but not sure if first point still applies

**Comment by memsharded** [comment]
Issue https://github.com/conan-io/conan/issues/3792 is also related to it: Conditional dependencies add a new level of options, that change the default IDs. Not sure if this should be changed, it could break all existing package IDs for all users, and the way to filter them is relatively understandable:

```python
from conans import ConanFile
class ConanLib(ConanFile):
    options = {"use_lib": [True, False]}
    default_options= "use_lib=False"
    def requirements(self):
        if self.options.use_lib:
            self.requires("optional/0.1@user/testing")
    def package_id(self):
        if self.options.use_lib:
            self.info.requires.remove("optional")
```

**Comment by memsharded** [comment]
Changing this from 2.0 to 1.10, at least to discuss it, because if there is a change in binary IDs, we need to address it asap, to enable the transition, otherwise 2.0 would break all existing binaries.

**Comment by lasote** [comment]
We still want to discuss it, right?

**Comment by lasote** [comment]
@memsharded moving again to 2.0 because we didn't discuss it yet and it is important to not lose it.
What do you think it could be done to not break anything? For the central repo, we might be able to generate different IDs for different modes, but for people in general...

**Comment by Minimonium** [comment]
Hey, folks. Is there a latest discussion on the matter anywhere?

**Comment by memsharded** [comment]
Not discussion, but we are doing a huge change towards 2.0 in https://github.com/conan-io/conan/pull/8996, as a result of the Tribe feedback (specifically the " the graph, including the build_requires is always expanded), and part of it is starting to explore the package_id limitations and new possibilities that the new graph brings, like the possibility of having the ``build_requires`` to affect the ``package_id`` as well. Once we have better understanding of the subject, we will continue discussing.

**Comment by Minimonium** [comment]
Alright, I'll leave a few notes for future reference:

* The current default package_id mode (semver_mode) uses an unrelated mode for indirect dependencies, which is especially problematic for Windows with `shared` option. Libraries that depend on such a library should be rebuilt (well, relink) when that option changes since they need to link to different `.lib` files. It's a frustrating experience _by default_, which is problematic for new users. Experienced users know to change the default mode to `recipe_revision_mode`.
* The issue is that conditional package id mode is only possible from consumers. Any such solution should prioritize _by-default_ robustness for newer users. Optimization should be an opt-in since it's more fragile to incompatible binary graphs.

**Comment by memsharded** [comment]
Hi @Minimonium 

Yes, totally, the default will definitely change in the new package_id. The work in progress in that #8996 PR is taking it into account already. It will be a changing value by default, changing depending on the shared/static/header, etc type of package. This will be default and automatic for consumers, nothing to do. Consider the following code already in that PR:

```python
          if dep_package_id is None:  # Automatically deducing package_id
                if require.include or require.link:  # linked
                    if node.package_type is PackageType.SHARED:
                        if dep_node.package_type is PackageType.SHARED:
                            dep_package_id = "minor_mode"
                        else:
                            dep_package_id = "recipe_revision_mode"
                    elif node.package_type is PackageType.STATIC:
                        if dep_node.package_type is PackageType.HEADER:
                            dep_package_id = "recipe_revision_mode"
                        else:
                            dep_package_id = "minor_mode"
                    elif node.package_type is PackageType.HEADER:
                        dep_package_id = "unrelated_mode"
```

This is nothing but a proof of concept, to understand what is possible, it doesn't try to be a proposal at all, just to show that this is possible.

**Comment by Minimonium** [comment]
The logic checks out to me! Looks great.

**Comment by memsharded** [comment]
This has been addressed already in 2.0 beta releases, explanations in this video too: https://www.youtube.com/watch?v=kKGglzm5ous
